lib/widgets/action_sheet.dart

```diff
@@ -1,36 +1,182 @@
+    // TODO(#24): Fix iOS bug where this call causes the keyboard to
+    //   reopen (if it was open at the time of this
+    //   `showMessageActionSheet` call) and cover a large part of the
+    //   share sheet.
+    Navigator.of(bottomSheetContext).pop();
+
+    // TODO: to support iPads, we're asked to give a
+    //   `sharePositionOrigin` param, or risk crashing / hanging:
+    //     https://pub.dev/packages/share_plus#ipad
+    //   Perhaps a wart in the API; discussion:
+    //     https://github.com/zulip/zulip-flutter/pull/12#discussion_r1130146231
+    // TODO: Share raw Markdown, not HTML
+    await Share.shareWithResult(message.content);
+  };
+}
+
+class QuoteAndReplyButton extends MessageActionSheetMenuItemButton {
+  QuoteAndReplyButton({
+    super.key,
+    required super.message,
+    required super.bottomSheetContext,
+    required super.messageListContext,
+  });
+
+  @override IconData get icon => Icons.format_quote_outlined;
+
+  @override String get label => 'Quote and reply';
+
+  @override VoidCallback get onPressed => () async {
+    // Close the message action sheet. We'll show the request progress
+    // in the compose-box content input with a "[Quoting…]" placeholder.
+    Navigator.of(bottomSheetContext).pop();
+
+    // This will be null only if the compose box disappeared after the
+    // message action sheet opened, and before "Quote and reply" was pressed.
+    // Currently a compose box can't ever disappear, so this is impossible.
+    ComposeBoxController? composeBoxController = MessageListPageState.composeBoxControllerOf(messageListContext);
+    final topicController = composeBoxController!.topicController;
+    if (
+      topicController != null
+      && topicController.textNormalized == kNoTopicTopic
+      && message is StreamMessage
+    ) {
+      topicController.value = TextEditingValue(text: message.subject);
+    }
+    final tag = composeBoxController.contentController
+      .registerQuoteAndReplyStart(PerAccountStoreWidget.of(messageListContext),
+        message: message,
+      );
+
+    Message? fetchedMessage;
+    String? errorMessage;
+    // TODO, supported by reusable code:
+    // - (?) Retry with backoff on plausibly transient errors.
+    // - If request(s) take(s) a long time, show snackbar with cancel
+    //   button, like "Still working on quote-and-reply…".
+    //   On final failure or success, auto-dismiss the snackbar.
+    try {
+      fetchedMessage = await getMessageCompat(PerAccountStoreWidget.of(messageListContext).connection,
+        messageId: message.id,
+        applyMarkdown: false,
+      );
+      if (fetchedMessage == null) {
+        errorMessage = 'That message does not seem to exist.';
+      }
+    } catch (e) {
+      switch (e) {
+        case ZulipApiException():
+          errorMessage = e.message;
+      }
+    }
+
+    if (fetchedMessage == null && errorMessage == null) {
+      errorMessage = 'Could not fetch message source.';
+    }
```

Comment:
Slightly easier to follow, I think, would be:
```suggestion
      switch (e) {
        case ZulipApiException():
          errorMessage = e.message;
        default:
          errorMessage = 'Could not fetch message source.';
      }
    }
```

Can then add an `assert(errorMessage != null)` inside the `if (fetchedMessage == null)` case below.